Revision based on feedback (#3)
mihaibudiu/sigmod-rh: relational.tex

%$$
%%\end{definition}

We define a handy function on \zrs called $\distinct$ (this function
is called $\epsilon$ in~\cite{griffin-sigmod95}, but we prefer our

> name because it is strongly related to the SQL \texttt{DISTINCT}

Reviewer: ```suggestion
name because it is closely related to the SQL \texttt{DISTINCT}
```

> operator).  This function, applied to a \zr, gives us another \zr

Reviewer: ```suggestion
operator).  This function, applied to a \zr, returns another \zr
```

%These properties allow us to ``consolidate'' distinct operators by performing
%one $\distinct$ at the end of a chain of computations.

\subsection{Incremental view maintenance}

Let us consider a relational query $Q$ defining a view $V$.  To create
a \dbsp circuit for $\inc{Q}$ that maintains incrementally $V$ we
apply the following mechanical steps:
%
\begin{algorithm}[incremental view maintenance]\label{algorithm-inc}\!
\begin{enumerate}
\item Translate $Q$ into a circuit using the rules in Table~\ref{tab:relational}.
\item{} [Optional] Remove some $\distinct$ operations.
\item Lift the whole circuit, converting it to a circuit operating
  on streams, using formula (**) in \refsec{sec:streams}.
\item Incrementalize the whole circuit ``surrounding'' it with $\I$ and $\D$.

> \item Apply the chain rule recursively on the circuit to obtain a

Reviewer: The last step really consists of applying the chain rule followed by "merging" D and I into the operator by replacing it with its incremental version where possible. This can probably be deduced from reading the paper carefully but I would make it very explicit

  circuit based only on primitive incremental operations.
\end{enumerate}
\end{algorithm}

This algorithm is deterministic and its running time is proportional
to the number of operators in the query.  Step (2) generates an
equivalent circuit, with possibly fewer $\distinct$ operators.  Step
(3) yields a circuit that consumes a \emph{stream} of complete
database snapshots and outputs a stream of complete view
snapshots. Step (4) yields a circuit that consumes a stream of
\emph{database changes} and outputs a stream of \emph{view changes};
however, the internal operation of the circuit is non-incremental, as
it rebuilds the complete database using integration operators.  Step
(5) optimizes the circuit by replacing each primitive operator with
its incremental version.

> After running this algorithm, all original primitive are replaced by

Reviewer: ```suggestion
After running this algorithm, all original primitives are replaced by
```

%$$
%\end{proposition}

The function $H$ has two inputs: the left input is the change $\Delta
d$, while the top input is the full set, obtained as an integral of
the changes.  $H$ detects whether the weight of a row in the full set
is changing sign (positive to negative or vice-versa) when the row
appears in a new change.  A precise definition of $H$ is given in the
full paper.  Here is the intuition why $\distinct$ is efficiently
incrementalizable: consider a row that is inserted in the input of a
$\distinct$ operator.  How does the output change?  If the row was not
in the prior full input table, then it will appear with weight 1 in

> the output change.  Otherwise an empty output is produced.  Only

Reviewer: should we explain the deletion too?

\subsection{IVM query plans and optimality}

> A relational algebra query can be implemented by multiple plans, each

Reviewer: I cannot comment on the SQL section above. Can we refer to a section in the full paper for group by and aggregation?
